### ISSUE TITLE
Better support for nested pipelines

### ISSUE THREAD

**Issue by shykes** [dagger/dagger#5303]
## Problem

Dagger has an API for logical grouping: the `Pipeline` step. Developers use it for nesting pipelines within other pipelines. But the `Pipeline` step relies on chaining semantics, not nesting. Because of this mismatch between the reality of how `Pipeline` works, and how developers *think* it works, everyone gets confused.

## Solution

I propose a new API for logically grouping, that supports nesting natively. As a result, what developers *think* will happen actually matches what *does* happen.

```golang
func example(ctx context.Context, dag dagger.Client) {
  dag.
    Step("build", func( dagger.Client) error {
      // Build step goes here
    }).
    Step("test", func( dagger.Client) error {
      // Test step goes here
    })
}
```

In summary: `Step` is exactly like `With`, but with a mandatory name argument, and a clearer name.

**Comment by shykes** [comment]
FYI @vito @helderco @sipsma @cpdeethree

**Comment by gerhard** [comment]
`Step()` suggests sequencing to me, e.g. run step 1, **then** step 2, etc. FWIW, this is how steps behave in GitHub Actions workflows.

I can see how `Job()` would be better suited for this logical grouping. The rest of the proposal makes sense to me 👍

While on this topic, I would like us to improve on the `Pipeline()` concept. BTW, there are three different scopes for `Pipeline()`:
1. https://pkg.go.dev/dagger.io/dagger#Client.Pipeline
2. https://pkg.go.dev/dagger.io/dagger#Container.Pipeline
3. https://pkg.go.dev/dagger.io/dagger#Directory.Pipeline

While I understand `Client.Pipeline()`, I am less sure about the other two.

If we had `Step()` - or `Job()` per my suggestion - would we deprecate & eventually remove `Pipeline()`?

**Comment by cpdeethree** [comment]
I also like Helder's comment on a [related PR](https://github.com/airbytehq/aircmd/pull/1#discussion_r1225285330) where I'm working on this  concept. Specifically around the idea that a Pipeline consists of a series of Steps, executed in sequence, and a Step itself can consist of either a single Pipeline or a list of Pipelines, and if a list of Pipelines is supplied, those Pipelines are executed concurrently

**Comment by shykes** [comment]
Thanks @cpdeethree and @kralicky. I can't tell if you're favor of this proposal, and if not, what change you're proposing? Is it about the choice of words ("step" vs "pipeline" vs something else) or about the functionality itself?

**Comment by gerhard** [comment]
> Steps are indeed sequenced.
> 
> `dag.Step(“lint”, …).Step(“build”, …).Step(“test”, …).Step(“deploy”, …)`
> 
> This runs the lint, build, test and deploy steps, in sequence.

OK, I am starting to get this.

The biggest disconnect  for me is that I think of `lint` as a pipeline, which runs in parallel to `build` & `test`. For Dagger itself, we run `lint`, `test` & `publish` in parallel.

The title suggests to me that this is meant to address pipeline nesting, but since a `Step` has different behaviour to `Pipeline`, perhaps a better title would be _Add grouping for pipeline operations_. It's not about nesting pipelines (which are parallel), but about **grouping operations** (which are already sequential, so step make sense).

Are you thinking of steps including other pipelines? This is an interesting take: https://github.com/airbytehq/aircmd/pull/1#discussion_r1221863715

**Comment by shykes** [comment]
I now regret introducing new terminology in this PR :) I will amend my proposal and add a section on terminology.

**Comment by cpdeethree** [comment]
After doing some work around this in my own pipeline code I can now give some more thorough feedback on this proposal.

One of the things that is currently being forced to be hand implemented by end users of Dagger right now is building reusable pipeline components and having a sane way to attach them to commands and run them in both sequence or parallel, passing previous steps to future ones. For me, whether you call it "step" or "job" or "pipeline" feels slightly aside to the main issue of how to collect individual commands ("build", "test", "lint", "publish") and execute them either individually as their own command (dagger do lint) or as part of a larger pipeline (dagger do ci, which runs build, test, lint, and publish, either in sequence or in parallel, as defined by the end user). We are currently forced to implement this orchestration ourselves.

The abstraction we ended up moving towards to be able to achieve this after some thought and conversation was decoupling the command from the task. [Here](https://github.com/airbytehq/aircmd/blob/95c4b765a1dc72d74a17fd1ea47bbb895809ce4f/aircmd/core/artifact.py) is a simple example that runs `ci` with two sub-pipelines, `build` and `test`. A Pipeline has a list of "steps" that is run when you execute the pipeline, and a "step" is just a raw function (leaf node) or another pipeline (container for nodes). You might have a `build` command that creates a `build` pipeline and executes that pipeline (build_task). `ci` command creates its own build pipeline that calls build_task. Then we have a top level executor (which is fulfilled by the role of `dagger do` in this case that recursively calls the pipelines until it reaches a leaf node (build_task).

Is the above proposal intended to address some of these issues around sequencing and orchestration?

**Comment by shykes** [comment]
@cpdeethree this proposal is a small, incremental change that removes a relatively well-scoped problem. Like you said, terminology isn't the important part thing here, and I will edit my proposal to make that more clear (didn't get around to it yet).

What you're describing is a broader DX discussion that can happen in parallel. It is covered by #4414, work by @sipsma on `dagger do`, etc. And yes, we do plan on expanding the scope of the Dagger API (and CLI) to address the problems you're describing - ideally with escape hatches for power users who don't want to buy into the whole thing (various tradeoffs there to be worked out).

Switching back to this proposal: it is a small, "chiropractic" adjustment to the low-level API, to solve a specific problem: logical grouping (via `Pipeline()`) and nesting (via `With()`) are two distinct features at the moment, but developers think of them interchangeably. Specifically, developers use `Pipeline()` expecting it to support nesting - but it doesn't. This causes confusion when visualizing the logical groups.

In this context of the Dagger engine API, the following terms are already defined:

* A *pipeline* is a linear sequence of steps encoded as a GraphQL query, which the Dagger Engine computes by modeling it as a DAG, and resolving it.
* A *step* is one discrete element of a pipeline. It is modeled by a single field in the GraphQL query (the leaf field is not considered a step). For example, the query `{ container { from("alpine"}) { withExec(args: ["echo", "hello world"]) { stdout } } } }` encodes a pipeline made of 3 steps: `container`, `from` and `withExec`.

From those two primitives, we can construct pretty much any other high-level concept: command, task, etc.
